04-onion-routing.md

```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
+   doesn't use the reply path.
+3. Mallory reuses the reply path later (perhaps after Alice has
+   forgotten about it), to reply to a different message.
+4. Mallory reuses the reply path to for a non-reply message.
+
+In all these cases, Alice must not reveal anything.  The requirements
+specify that for (3) and (4) the message is ignored, and that for (2)
+Alice responds however she would if she didn't expect the reply
+(usually ignore, or error).
+
+Similarly, if Alice publishes a blinded path to reach her node, and
+Mallory sends a reply to another message from Alice, using the blinded
+path.  In this case, Alice ignores the message.
+
+This raises the question of how Alice knows whether a message was sent
+via a path she supplied.  One method is to remember the final
+`blinding` expected, but requires remembering every such blinding.
+This is the purpose of the `self_id` field: Alice can place a secret
+inside this field, knowing only she can read it, and thus verify the
+path as one she created.  It is variable length so she can also
+add data to identify which path it was.
+
+#### Requirements
+
+The writer:
+- For the non-final nodes' `onionmsg_payload`:
+  - MUST set `enctlv` to a valid `encmsg_tlvs` containing `next_node_id`.
+  - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].
+    `ChaChaPoly-1305` encryption using an all-zero nonce).
+  - MAY include `padding`.
+  - MUST NOT set `self_id`.
+- For the final node's `onionmsg_payload`:
+  - if the final node is permitted to reply:
+    - MUST set `reply_path` `blinding` to the initial blinding factor for the `next_node_id`
+    - MUST set `reply_path` `first_node_id` to the unblinded node id of the first node in the reply path.
+    - For every `reply_path` `path`:
+      - MUST set `node_id` to the blinded node id to encrypt the onion hop for.
+      - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].
+        `ChaChaPoly-1305` encryption using an all-zero nonce).
+      - MUST set `enctlv` to a valid `encmsg_tlvs` which meets the requirements
+        of the `onionmsg_payload` when used by the recipient.
+      - MAY use `self_id` to contain a secret so it can recognize use of this `reply_path`.
+  - otherwise:
+    - MUST NOT set `reply_path`.
+
+The reader:
+- if it is not the final node according to the onion encryption:
+  - if `enctlv` is not present, or does not decrypt with the shared secret from the given `blinding` parameter:
+    - MUST drop the message.
+  - if the `enctlv` is not a valid TLV or does not contain `next_node_id`:
+    - MUST drop the message.
+  - if the `enctlv` contains `self_id`:
+    - MUST drop the message.
+  - otherwise:
+    - MUST ignore `padding`, if any.
+    - SHOULD forward the message using `onion_message` to the next peer indicated by `next_node_id`.
+    - if it forwards the message:
+      - if `blinding` is specified in the `enctlv`:
+         - MUST pass that as `blinding` in the `onion_message`
```

Comment:
Why not always use the derived `blinding`? What is the use-case where we need to override the default `blinding`?

Reply:
If I give you a blinded path from Alice, you need to make an onion which goes to Alice then uses the blinded path.  This requires blinding override.

Reply:
Can you detail that @rustyrussell? I don't understand it either.
If for example Carol creates a blinded path `Alice -> Bob -> Carol`.
Carol provides a list of `encrypted_recipient_data` and a first `blinding` point that must be sent to Alice.
But that `blinding` cannot be inside the `encrypted_recipient_data` as it's required to actually decrypt it...?
It seems to me that you just need the `blinding` tlv fields that is part of the `onion_message` for that.

Reply:
I receive the following blinded route built by Carol: `(blinding0, Alice, (BlindedAlice, BlindedAliceData) -> (BlindedBob, BlindedBobData) -> (BlindedCarol, BlindedCarolData))` and I want use it by adding Dave in front.

I want to create a route `(blinding1, Dave, (BlindedDave, BlindedDaveData) -> (BlindedAlice, BlindedAliceData) -> (BlindedBob, BlindedBobData) -> (BlindedCarol, BlindedCarolData))`. For that I generate randomly `blinding1` and use it to blind `(Dave, DaveData)`, I then obtain `(BlindedDave, BlindedDaveData)` and a derived `blinding1Next`. However `blinding1Next` is not equal to `blinding0`, so I need to specify `blinding0` as an override for the rest of the route. I pass this `blinding0` to Dave, inside `DaveData` and Dave MUST pass this `blinding0` as `blinding` in the `onion_message` it sends to Alice.

Reply:
If you're concatenating two blinded routes I can see why you would need that, but is there a use-case for that though?
In all scenarios I can think of you would use a normal route through Dave, the blinded part would only start at Alice.
The recipient is the one that wants to hide, why would the sender would use a blinded route for the part that he chooses?

Reply:
There is no such thing as a normal route. Everything is blinded.

Reply:
Right, that makes sense! It's worth explaining a bit more in the PR then, the rationale section should make this obvious, since route blinding is designed to prepend a normal route to a blinded one (at least for payment scenarios).